litmus-portal/graphql-server/pkg/chaoshub/ops/gitops_test.go

```diff
@@ -0,0 +1,203 @@
+package chaoshubops_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/gin-gonic/gin"
+	"github.com/google/uuid"
+	"github.com/litmuschaos/litmus/litmus-portal/graphql-server/graph/model"
+	chaosHubOps "github.com/litmuschaos/litmus/litmus-portal/graphql-server/pkg/chaoshub/ops"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+)
+
+// TestMain is the entry point for testing
+func TestMain(m *testing.M) {
+	gin.SetMode(gin.TestMode)
+	log.SetOutput(ioutil.Discard)
+	os.Exit(m.Run())
+}
+
+// clearCloneRepository removes the cloned repository
+func clearCloneRepository(projectID string) {
+	tempPath := chaosHubOps.GetClonePath(chaosHubOps.GitConfigConstruct(model.CloningInput{ProjectID: projectID}))
+	err := os.RemoveAll(tempPath)
+	if err != nil {
+		panic(fmt.Sprintf("failed to remove temp path: %v", err))
+	}
+}
+
+// TestGetClonePath is used to test the GetClonePath function
+func TestGetClonePath(t *testing.T) {
+	// given
+	projectID := uuid.New().String()
+	chaosHubConfig := chaosHubOps.ChaosHubConfig{
+		ProjectID: projectID,
+		HubName:   "test",
+	}
+	// when
+	path := chaosHubOps.GetClonePath(chaosHubConfig)
+	// then
+	assert.Equal(t, "/tmp/version/"+projectID+"/test", path)
+}
+
+// TestGitConfigConstruct is used to test the GitConfigConstruct function
+func TestGitConfigConstruct(t *testing.T) {
+	// given
+	repoData := model.CloningInput{
+		ProjectID: uuid.New().String(),
+		HubName:   "test",
+		RepoURL:   "url",
+	}
+	// when
+	config := chaosHubOps.GitConfigConstruct(repoData)
+	// then
+	assert.Equal(t, repoData.ProjectID, config.ProjectID)
+	assert.Equal(t, repoData.HubName, config.HubName)
+	assert.Equal(t, repoData.RepoURL, config.RepositoryURL)
+}
+
+// TestGitClone is used to test the GitClone function
+func TestGitClone(t *testing.T) {
+	// given
+	testcases := []struct {
+		name     string
+		repoData model.CloningInput
+		isError  bool
+	}{
+		{
+			name: "success",
+			repoData: model.CloningInput{
+				ProjectID:  uuid.New().String(),
+				HubName:    "hub1",
+				RepoURL:    "https://github.com/litmuschaos/chaos-charts",
+				RepoBranch: "master",
+				IsPrivate:  false,
```

Comment:
Can you add private=true test cases for each of the test types (Sync, clone, pull, etc) too! In the same handler and assert it too.

Reply:
It's cool but, our current unit test runs an actual git clone job so we need a real private repository and private key to run this job. IMO It's not reasonable for testing with a real private key in the unit test. if we want to test private repository test cases, we need to mock the git clone parts (make a new interface to run GitClone) and not run the actual git clone. What do you think?

Reply:
@namkyu1999 maybe we can do a negative test with random creds and make sure we get an unauthorized response, because at the end of the day the git clone is handled by go-git and we don't have to test that what we need to test is whether we are calling the go-git package with proper data and imo if we can even get an unauthorized response for a valid private git repo we should be good. 

Also if we absolutely want to test private repos we can add a secret to the repo and then use it as an env for the test to read. Though this may have security concerns, we can minimize it to some extent by limiting the access of the test creds to only 1 dummy repo

Reply:
Okay, then I will make that test cases too!

Reply:
Yes , @namkyu1999  you can add the negative test cases for the git functionalities for now. And for the positive test cases that @gdsoumya mentioned, we can take it in a different PR. We can break these test case PRs in different chunks for more flexibility in tracking.

Reply:
@S-ayanide , I added negative test cases and i will add positive test cases in different pr just like @amityt mentioned.

Reply:
Awesome, +1 to @gdsoumya's points, asserting the error response would also count as valid in this case. That could be the first approach we take.